logrus/logrus.go

```diff
@@ -72,27 +97,87 @@ func (rl *Logger) Fatal() slog.Logger {
 
 // WithLevel returns a new logger set to add entries to the specified level
 func (rl *Logger) WithLevel(level slog.LogLevel) slog.Logger {
-	return rl
+	var levels = []logrus.Level{
+		slog.UndefinedLevel: logrus.TraceLevel + 1,
+		slog.Fatal:          logrus.FatalLevel,
+		slog.Error:          logrus.ErrorLevel,
+		slog.Warn:           logrus.WarnLevel,
+		slog.Info:           logrus.InfoLevel,
+		slog.Debug:          logrus.DebugLevel,
+	}
+
+	if level < slog.UndefinedLevel || int(level) >= len(levels) {
+		// fix your code
+		err := fmt.Errorf("slog: invalid log level %v", level)
+		rl.logger.Panic(err)
+	}
+
+	out := rl.dup(nil)
+	out.level = levels[level]
+	return out
 }
 
-// WithStack attaches a call stack to a new logger
+// WithStack attaches a call stack to the log entry
 func (rl *Logger) WithStack(skip int) slog.Logger {
+	if rl.Enabled() {
+		frames := internal.StackTrace(skip + 1)
+		if len(frames) > 0 {
+			caller := frames[0]
+
+			entry := rl.entry.WithFields(logrus.Fields{
+				CallerFieldName: fmt.Sprintf("%+n", caller),
+				StackFieldName:  fmt.Sprintf("%+n", frames),
+			})
+
+			return rl.dup(entry)
+		}
+	}
 	return rl
 }
 
-// WithField returns a new logger with a field attached
+// WithField adds a field to the log entry
 func (rl *Logger) WithField(label string, value any) slog.Logger {
+	if rl.Enabled() {
+		entry := rl.entry.WithFields(logrus.Fields{
+			label: value,
+		})
+		return rl.dup(entry)
+	}
 	return rl
 }
 
-// WithFields returns a new logger with a set of fields attached
+// WithFields adds fields to the log entry
 func (rl *Logger) WithFields(fields map[string]any) slog.Logger {
+	if rl.Enabled() {
+		entry := rl.entry.WithFields(fields)
+		return rl.dup(entry)
+	}
 	return rl
 }
 
-// NewWrapper creates a slog.Logger adaptor using a logrus as backend
-func NewWrapper(entry *logrus.Logger) slog.Logger {
+// New creates a slog.Logger adaptor using a logrus as backend
+func New(logger *logrus.Logger) slog.Logger {
+	if logger == nil {
+		return nil
+	}
+
+	return &Logger{
+		logger: logger,
+		entry:  logrus.NewEntry(logger),
+		level:  logrus.InfoLevel,
```

Comment:
You are changing here the level of the received logger to Info. Why?

Reply:
I hadn't realised there was a received level 😅

Reply:
logger.GetLevel()

Reply:
I thought that was linked to `IsLevelEnabled()`

Reply:
yes, according to https://pkg.go.dev/github.com/sirupsen/logrus#Logger.IsLevelEnabled the Level of the logger is the threshold to decide if a given level is enabled or not

Reply:
Well.... in a sense the level of the Logger is the minimum level that will be logged so imho
```go
logger.GetLevel()
```
is what we need here and not the default 
```go 
InfoLevel
```